src/Microsoft.AspNetCore.Server.Kestrel/Filter/StreamSocketOutput.cs

```diff
@@ -65,14 +71,50 @@ public void ProducingComplete(MemoryPoolIterator end)
             var block = _producingBlock;
             while (block != end.Block)
             {
-                _outputStream.Write(block.Data.Array, block.Data.Offset, block.Data.Count);
+                // Writes can fail if the client has closed the connection but we're still sending a response.
+                // If we don't handle an exception from _outputStream.Write() here, we'll leak memory blocks.
+                if (_canWrite)
+                {
+                    try
+                    {
+                         _outputStream.Write(block.Data.Array, block.Data.Offset, block.Data.Count);
+                    }
+                    catch (ObjectDisposedException ex)
+                    {
+                         _canWrite = false;
+                         _logger.ConnectionDisconnectedWrite(_connectionId, block.Data.Count, ex);
+                    }
+                    catch (Exception ex)
+                    {
+                        _canWrite = false;
+                        _logger.ConnectionError(_connectionId, ex);
+                    }
+                }
 
                 var returnBlock = block;
                 block = block.Next;
                 returnBlock.Pool.Return(returnBlock);
             }
+            
+            if (_canWrite)
+            {
+                var count = end.Index - end.Block.Data.Offset;
```

Comment:
Nit: this can be moved inside of the `try` once you only log using only `ConnectionError`